internal/mempool/mempool.go

```diff
@@ -2159,4 +2195,457 @@ func New(cfg *Config) *TxPool {
+// findTx.
+func (mv *txMiningView) addTransaction(txDesc *mining.TxDesc, findTx txDescFind) {
+	// Track the txn in the graph and attempt to relate it to at least one other
+	// transaction.
+	mv.txGraph.insert(txDesc, findTx)
+
+	// Ensure that the statistics for this transaction are up to date
+	// given the current ancestors in the view.
+	mv.updateBundleStats(txDesc)
+
+	// If a transaction is added back to the view,
+	// reconnect it with transactions already in the view.
+	// This is necessary to account for reorgs.
+	mv.updateStatsDescendantsAdded(txDesc)
+}
+
+// TxDescs returns a collection of all transactions available in the view.
+func (mv *txMiningView) TxDescs() []*mining.TxDesc {
+	return mv.txDescs
+}
+
+// findTx returns a transaction from the mempool by hash.  If it does not exist
+// in the mempool, a nil pointer is returned.
+func (mp *TxPool) findTx(txHash *chainhash.Hash) *mining.TxDesc {
+	poolTx := mp.pool[*txHash]
+	if poolTx == nil {
+		return nil
+	}
+
+	return &poolTx.TxDesc
+}
+
+// find returns the TxDesc stored in the graph by its hash. If the transaction
+// is not in the graph, then a nil pointer is returned. Since each transaction
+// in the graph must have at least one edge connecting it to another
+// transaction, scanning for a known hash as a child or parent is sufficient
+// to recover its pointer.
+func (g *txDescGraph) find(txHash *chainhash.Hash) *mining.TxDesc {
+	for parentHash := range g.parentsOf[*txHash] {
+		return g.childrenOf[parentHash][*txHash]
+	}
+
+	for childHash := range g.childrenOf[*txHash] {
+		return g.parentsOf[childHash][*txHash]
+	}
+
+	return nil
+}
+
+// aggregateAncestorStats adds the provided ancestor transaction's statistics
+// to the provided ancestor stats reference.
+func aggregateAncestorStats(stats *mining.TxAncestorStats, ancestor *mining.TxDesc) {
+	stats.NumAncestors++
+	stats.Fees += ancestor.Fee
+	stats.SizeBytes += int64(ancestor.Tx.MsgTx().SerializeSize())
+	stats.NumSigOps += int64(ancestor.NumSigOps)
+	stats.NumP2SHSigOps += int64(ancestor.NumP2SHSigOps)
+}
+
+// updateBundleStats ensures that the bundle stats for a given
+// transaction are up to date. It aggregates statistics for the provided
+// transaction based on its ancestors.
+func (mv *txMiningView) updateBundleStats(txDesc *mining.TxDesc) {
+	seen := make(map[chainhash.Hash]struct{})
+	txHash := txDesc.Tx.Hash()
+	stats := &mining.TxAncestorStats{}
+	mv.txGraph.forEachAncestor(txHash, seen, func(txDesc *mining.TxDesc) {
+		aggregateAncestorStats(stats, txDesc)
+	})
+
+	mv.ancestorStats[*txDesc.Tx.Hash()] = stats
+}
+
+// clone returns a copy of the current graph.
+func (g *txDescGraph) clone(fetchTx txDescFind) *txDescGraph {
+	// Source transactions from within the graph to decouple the cloned
+	// mining view instance from the mempool.
+	forEachRedeemer := func(tx *dcrutil.Tx, f func(redeemerTx *mining.TxDesc)) {
+		for _, childTx := range g.childrenOf[*tx.Hash()] {
```

Comment:
Updated, good catch.

forEachRedeemer is only called when inserting into the view, so the behavior was the same since the TxMiningView interface is readonly.